File: src/main/scala/com/github/shadowsocks/plugin/obfs_local/BinaryProvider.scala
package com.github.shadowsocks.plugin.obfs_local

import java.io.{File, FileNotFoundException}

import android.net.Uri
import android.os.ParcelFileDescriptor
import com.github.shadowsocks.plugin.{NativePluginProvider, PathProvider}

/**
  * @author Mygod
  */
final class BinaryProvider extends NativePluginProvider {
  override protected def populateFiles(provider: PathProvider): Unit = provider.addPath("obfs-local", "755")

  override def getExecutable: String = getContext.getApplicationInfo.nativeLibraryDir + "/libobfs-local.so"

  override def openFile(uri: Uri): ParcelFileDescriptor = uri.getPath match {
    case "/obfs-local" => ParcelFileDescriptor.open(new File(getExecutable), ParcelFileDescriptor.MODE_READ_ONLY)
    case _ => throw new FileNotFoundException()
  }
}

File: app/src/main/kotlin/com/github/shadowsocks/plugin/obfs_local/BinaryProvider.kt
package com.github.shadowsocks.plugin.obfs_local

import android.net.Uri
import android.os.ParcelFileDescriptor
import com.github.shadowsocks.plugin.NativePluginProvider
import com.github.shadowsocks.plugin.PathProvider
import java.io.File
import java.io.FileNotFoundException

/**
 * @author Mygod
 */
class BinaryProvider : NativePluginProvider() {
    override fun populateFiles(provider: PathProvider) {
        provider.addPath("obfs-local", "755")
    }

    override fun getExecutable(): String {
        return context!!.applicationInfo.nativeLibraryDir + "/libobfs-local.so"
    }

    override fun openFile(uri: Uri, mode: String): ParcelFileDescriptor? {
        return when (uri.path) {
            "/obfs-local" -> ParcelFileDescriptor.open(
                File(getExecutable()),
                ParcelFileDescriptor.MODE_READ_ONLY
            )
            else -> throw FileNotFoundException()
        }
    }
}
